client/src/component/AddVideo.jsx

```diff
@@ -0,0 +1,166 @@
+      }
+    } catch (error) {
+      console.error("Error fetching videos:", error);
+    }
+  };
+
+  const handleInputChange = (event) => {
+    const { name, value } = event.target;
+    setNewVideo({ ...newVideo, [name]: value });
+  };
+
+  const getYouTubeVideoId = (url) => {
+    const videoIdMatch = url.match(/(?:\/|v=)([A-Za-z0-9_-]{11})(?=&|$)/);
+    if (videoIdMatch) {
+      return videoIdMatch[1];
+    }
+    return null;
+  };
+
+  const addVideo = async () => {
+    if (newVideo.title && newVideo.url) {
+      const videoId = getYouTubeVideoId(newVideo.url);
+
+      if (videoId) {
+        const videoToAdd = {
+          ...newVideo,
+          rating: 0,
+          id: idCounter,
+          timestamp: new Date().toISOString(),
+        };
+
+        try {
+          const response = await fetch(
+            "https://node-server-full-stack.onrender.com/videos",
+            {
+              method: "POST",
+              headers: {
+                "Content-Type": "application/json",
+              },
+              body: JSON.stringify(videoToAdd),
+            }
+          );
+
+          if (response.ok) {
+            // const responseData = await response.json();
+            // setVideos([...videos, responseData]);
+            setNewVideo({ title: "", url: "" });
+            setIdCounter(idCounter + 1);
+          } else {
+            console.error("Failed to add video:", response.statusText);
+          }
+        } catch (error) {
+          console.error("Error adding video:", error);
+        }
+      } else {
+        alert("Invalid YouTube URL. Please enter a valid YouTube video URL.");
+      }
+    }
+  };
+
+  const upvoteVideo = (id) => {
+    const updatedVideos = videos.map((video) =>
+      video.id === id ? { ...video, rating: video.rating + 1 } : video
+    );
+    setVideos(updatedVideos);
+    <p>{console.log("line 211", updatedVideos)}</p>;
+  };
+
+  const downvoteVideo = (id) => {
+    const updatedVideos = videos.map((video) =>
+      video.id === id && video.rating > 0
+        ? { ...video, rating: video.rating - 1 }
+        : video
+    );
+    <p>{console.log("line 220", updatedVideos)}</p>;
+
+    setVideos(updatedVideos);
+  };
+
+  const removeVideo = (id) => {
+    const updatedVideos = videos.filter((video) => video.id !== id);
+    setVideos(updatedVideos);
+  };
+
+  return (
+    <>
+      <div>
+        <h2>Add a Video</h2>
+        <input
+          type="text"
+          name="title"
+          placeholder="Title"
+          value={newVideo.title}
+          onChange={handleInputChange}
+        />
+        <input
+          type="text"
+          name="url"
+          placeholder="URL"
+          value={newVideo.url}
+          onChange={handleInputChange}
+        />
+        <button onClick={addVideo}>Add Video</button>
+      </div>
+      <div>
+        <h2>Videos</h2>
+        <ul className="ShowingVideos">
+          {videos.map((video) => (
```

Comment:
It would probably be more clear to extract a component for "showing one video", rather than having to work out what exactly is happening inside the `map` callback.